KEP-4815 DRA Partitionable devices design update

mortent opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: Update the API of the feature to use the abstract pools design.

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/4815

<!-- other comments or additional information -->
- Other comments: This PR also moves this feature from sig-node to sig-scheduling.

mortent commented:
/assign @pohly 
/assign @klueska 
/assign @johnbelamaric

johnbelamaric commented:
/lgtm

This change will ensure we get *some* implementation merged in 1.33. That said, I would like to explore a revision that moves back to the "abstract resource pools" model Kevin proposed very early on and is similar to those discussed [here](https://github.com/kubernetes-sigs/wg-device-management/issues/20#issuecomment-2168189769) and [here](https://github.com/kubernetes-sigs/wg-device-management/pull/27).

In particular, using that model instead will remove the confusion which we are working around in this KEP change by limiting `ConsumesCapacityFrom` to a single entry. Instead, every entry will be explicit as to which bucket its resources come from. More importantly, the current KEP assumes the advertised capacity of the device is identical to the capacity drawn down from the parent device. I don't think that will always be true. For example, in a different technology, there could be sharing overhead. So, you would consume 10GB but only advertise 9GB. Even in this example, there is no real need to advertise "memorySlice4" as a capacity in a MIG for end users. An abstract pool model would separate the "internal shared resources" from the "end user advertised resources".

pohly commented on keps/sig-node/4815-dra-partitionable-devices/README.md:
```diff
@@ -647,7 +655,12 @@ When such a device is allocated, the scheduler will need to track the full
+device it references directly.
+
+The API defines the `ConsumesCapacityFrom` field as a list of `DeviceRef` entries.
+While we only allow a single entry in this list, effectily forcing the partitioning
```
Nit: 
```suggestion
While we only allow a single entry in this list, effectively forcing the partitioning
```

mortent commented on keps/sig-node/4815-dra-partitionable-devices/README.md:
```diff
@@ -267,12 +267,16 @@ The basic idea is the following:
    objects directly next to the list of `Devices` inside a `ResourceSlice`.
-   They are not allocatable on their own.
+   They are not allocatable on their own. Entries in the `Includes` list can
+   reference mixins in any `ResourceSlice` in the same pool. References to
```
@pohly I added this clarification as it seems logical to me and also helpful with large number of devices that might need to be split across multiple ResourceSlices. But it also means that we can't apply mixins to devices as part of conversion since there can be cross-ResourceSlice references. Do you have any opinion on whether we should allow this or keep it simpler by restricting it to just within ResourceSlice references?

pohly replied:
I prefer to keep it simpler by restricting it to just within ResourceSlice references.

That enables validation of mixin references (validation is per slice, not per pool) and a potential performance enhancement (expanding references once [when receiving a slice](https://github.com/pohly/kubernetes/commit/cb18e0f16094d7b669bfeb8207de48705dc046ce)).

mortent replied:
Yeah, agree that validation will be better if we restrict references to be within a single ResourceSlice. Also, it is easy to loosen this restriction later if we decide it is needed. I will update the language here.

mortent replied:
Updated the language to clarify that references can only be internal to a ResourceSlice.

mortent commented:
@klueska This PR updates the KEP according to the discussion in https://kubernetes.slack.com/archives/C0409NGC1TK/p1737092728963179 where we discussed limiting the number of entries in the `consumesCapacityFrom` field to one. It also makes a few minor clarifications. Could you take a look?

mortent commented:
This change to the design restricts the number of `consumesCapacityFrom` references allowed for a single device to 1. In practice this means that a partition can only be part of a single source, or in other words, the hierarchy of partitions for a single top-level device must form a tree.

This works for several of the use-cases we have looked at, including:
* Partitioning NVIDIA Multi-Instance GPUs into smaller partitions (described in the original KEP)
* Multi-host scheduling of interconnected TPUs (described in https://github.com/kubernetes/enhancements/pull/5069)
* SR-IOV (described in https://github.com/byako/enhancements/commit/199d51362e19a969d6cf71a3b88e62825fda9abe)

However, it is not sufficient for some other examples, including single-host TPUs: https://docs.google.com/document/d/1lXGfnrBixRIMW9ESa-mv09Kisb2myVFV_A3nqPJ4FCQ/edit?tab=t.0#bookmark=id.k34x2m6g3rgo. Since this requires that a single TPU can be part of more than one "group" to TPUs, each TPU requires two entries in the `consumesCapacityFrom` list. I think this is also true for AWS accelerators: https://docs.google.com/document/d/1lXGfnrBixRIMW9ESa-mv09Kisb2myVFV_A3nqPJ4FCQ/edit?tab=t.0#bookmark=id.l90rxj27wqvp. 

The original design on this KEP does allow for multiple source devices (i.e. entries in the `consumesCapacityFrom` list) and allows for a partition to consume just part of the capacity of the source device. This requires that for every allocated device, the ResourceClaim (probably) needs to record exactly how much capacity was used from each source device (which might be 0 in some cases). And it needs to capture this for possibly multiple layers of source devices.

A possible "middle ground" between the original proposal and the one suggested in this PR, is to allow for multiple source devices, but not allow "partial use" of those source devices. This would let us support the use-cases above, but without the need to capture additional information about an allocation in the ResourceClaim. In this case, a device can be allocated if two conditions are met:
* All devices in the subtree underneath it are allocatable.
* All devices in all possible paths to the final source device is allocatable when following the `consumesCapacityFrom` references.

johnbelamaric commented:
@mortent yes, I think disallowing partial consumption makes it work. I still think we should look at abstract pools, because I suspect we will *need* partial consumption; I also don't think it makes sense to automatically publish all the resources we consume, which I *think* this current design does. That is, internally consumable by a partition is not the same as externally consumable by a claim, and I think we should maintain those as separate things. That's more-or-less the heart of the abstract pools proposal.

mortent commented:
@klueska If we only allow a single source device, then partial consumption is kinda irrelevant from the perspective of allocation. If device A has a `consumesCapacityFrom` reference to device B, then device A can only be allocated if device B is also allocatable. And once device A is allocated, device B can not be allocated regardless of how much capacity device A consumes (just having the reference means it consumes something, which makes device B unavailable).

So the question was mostly about whether partial consumption is a requirement or not. As far as I know, we don't have any use-cases for this feature yet. But from the comment from @johnbelamaric, it sounds like we think this is an important feature. I'm also a bit uncertain about the ability to add partial consumption later if we went with a solution for now that didn't include it.

I'm planning to try out the use-cases we have with the abstract pool feature. As you also mentioned on Slack, I think we will run into the challenge of tracking partial usage with that approach too, but we will see how it turns out.

klueska commented:
You mean partially consumed, but still available? Yes, I wouldn't imagine that to be supported (wheter we allow consumption from multiple source devices or not). But we definitely need to be able to allow multiple "sink-devices" to pull part of the capacity from the same "source-device" (and not just all of it).

mortent commented:
Yeah, my comment above isn't correct. In the "only one source" scenario, a device can only be allocated if all of its source devices (all the way to the root source) are not allocated (they don't have to be allocatable). But I think it is still correct that we can handle the allocation process correctly in this situation without actually tracking partial consumption. But the language in my original comment is not precise enough. I think the challenge arises when the use of multiple sources introduces ambiguity about where the capacity is consumed from. But I need to think a bit more about this.

mortent commented:
/wg device-management

klueska commented:
Yes, this is how I understood the problem since you first described it (which is why limiting things to one source would be one potential way to solve it).

Here is a reference to my comment in the slack conversation around this for reference:
https://kubernetes.slack.com/archives/C0409NGC1TK/p1737102143861779?thread_ts=1737092728.963179&cid=C0409NGC1TK

There have been a few ideas around how the abstract pools would actually look. But I'm hoping they take shape such that *all* devices are actually sink-devices that must pull their capacity from (at least one) abstract pool. I also *strongly* dislike the idea of "inheriting" attributes and/or capacities (which I think at least one of the abstract pool models allowed), and would like to see the concept of mixins persist in place of that.

mortent commented:
@johnbelamaric @klueska I've updated the proposed API changes in the [Design Details](https://github.com/kubernetes/enhancements/blob/317b38cea8862b337d6c61e7ed6802184a6774ae/keps/sig-node/4815-dra-partitionable-devices/README.md#design-details) section as well as the small examples in the [Extending a device with a set of mixins](https://github.com/kubernetes/enhancements/blob/317b38cea8862b337d6c61e7ed6802184a6774ae/keps/sig-node/4815-dra-partitionable-devices/README.md#extending-a-device-with-as-set-of-mixins) and [Defining device partitions in terms of consumed capacity in a composite device](https://github.com/kubernetes/enhancements/blob/317b38cea8862b337d6c61e7ed6802184a6774ae/keps/sig-node/4815-dra-partitionable-devices/README.md#defining-device-partitions-in-terms-of-consumed-capacity-in-a-composite-device) sections.

Could you take a look and those and give feedback? I know there has been discussions before around this that I wasn't part of. I've primarily looked at the structure and naming from https://github.com/kubernetes-sigs/wg-device-management/pull/27.
I want to see if we can agree on the details here as early as possible, at least before I start reworking the large examples.

johnbelamaric commented:
> Could you take a look and those and give feedback? 

Just noting here what you and I discussed in person. I think this is directionally correct, a few notes:

1. As Kevin mentioned, I think only pools should be sources; all devices are sinks.
2. I think we can have multiple pools, not just a single pool the way it is in your current example.
3. Taking the mixins a step further, I'd like to see mixins that represent the pool capacity, and mixins that represent the pattern of consumption. To consume, a device would specify the pool(s) from which they consume, and what they consume from those pools. To list the "what they consume", they could list resources directly, or pull in a mixin that represents the consumption pattern.

With that, we can make it pretty compact while being pretty understandable. For example, we would have something like:

```yaml
sharedCapacityMixins:
- name: a100-capacity
  capacity:
     memory: 40Gi
     copy-engines: "7"
   
sharedCapacityPools:
- name: gpu-0
  mixins: ["a100-capacity"]
  capacity:
    something-not-in-the-mixin: "872"

sharedCapacityConsumptionMixins:
- name: mig-1g.5gb
   consumesCapacity:
     memory: 5Gi
     copy-engines: 1
 
devices:
- name: gpu-0-mig-1g.5gb-0
  consumesCapacity:
  - capacityPool: gpu-0
     mixins: ["mig-1g.5gb"]
     capacity:
       something-not-in-the-mixin: "2"
  providesCapacity:
    mixins: ["a mixin representing this mig profile advertised capacity"]
    capacity:
     ...list of resources outside of mixins...
```

In a typical case, I wouldn't expect any capacities outside of the mixins (maybe we even leave that out for now?). So, each device becomes just few lines:
* name
* list of attribute mixins
* list of device specific attributes
* list of pools, and the mixins consumed from that pool
* list of capacity mixins (maybe attribute & capacity mixins can be combined?)

mortent commented:
Thanks @johnbelamaric 

I added another commit about the same time as you added your comment with some changes that adds a proposal for what we discussed in person. It is mostly similar I think, but some interesting differences:

Your example with the proposal I added will look something like this:
```yaml
sharedCapacityPools:
- name: gpu-0
  capacities:
  - name: memory
    capacity: 40Gi
  - name: copy-engines
    capacity: 7
  - name: something-not-in-the-mixin
    capacity: 872
mixins:
- name: mig-1g.5gb
  composite:
    capacity:
      number-of-foos: 14
    sharedCapacityConsumed:
    - name: memory
      capacity: 5Gi
    - name: copy-engines
      capacity: 1    
devices:
- name: gpu-0-mig-1g.5gb-0
  composite:
    includes:
    - name: mig-1g.5gb
      capacityPool: gpu-0
    sharedCapacityConsumed:
      something-not-in-the-mixin: "2"
    capacity:
      number-of-bars: 42
```

A few differences:
* It doesn't have support for mixins in the `sharedCapacityPool`. That does seem pretty useful, although we end up with multiple layers of mixins that might make it hard to follow.
* It only uses a single type of mixin for the device, which includes `Attributes` and `Capacity` as today, but also adds `SharedCapacityConsumed`. I'm a bit uncertain if I prefer a single mixin for all three or have separate mixins for consumed capacity and advertised capacity.
* A few things I want to clean up in my proposal like inconsistent use of map vs list-with-name-field and it also still has the `composite` field.

mortent commented:
I took another pass at this and made some changes to the design. 

Most importantly, I added support for capacity pool mixins and consistently use `Capacity map[QualifiedName]DeviceCapacity` for specifying capacity across devices, capacity pools and capacity pool mixins. I left the composite device in there, since I don't think we have fully decided on the path forward there, it is a relatively simple thing to remove, and we probably need it in the `v1beta1` API anyway. 

I also kept the "list-with-name-field" pattern for all references. It is a little more verbose than using a map, but it gives us the flexibility to add additional fields in the future.

The example then becomes:
```yaml
capacityPoolMixins:
- name: a100-capacity
   capacity:
      memory: 40Gi
      copy-engines: "7"
sharedCapacityPools:
- name: gpu-0
  includes:
  - name: a100-capacity
  capacity:
    something-not-in-the-mixin: "872"
mixins:
- name: mig-1g.5gb
  composite:
    capacity:
      number-of-foos: "14"
    sharedCapacityConsumed:
    - name: memory
      capacity: 5Gi
    - name: copy-engines
      capacity: "1"
devices:
- name: gpu-0-mig-1g.5gb-0
  composite:
    includes:
    - name: mig-1g.5gb
      capacityPool: gpu-0
    sharedCapacityConsumed:
    - name: something-not-in-the-mixin
      capacity: "2"
    capacity:
      number-of-bars: "42"
```

Overall I think I like the structure, as it gets pretty compact for most scenarios.

I see at least two things that we could change to possibly improve the API:
* Remove the structs for the refs, and make them either just a list of strings (example: `sharedCapacityPools.includes`) or a map (example: `mixins.composite.sharedCapacityConsumed`).
* Create a separate variant of the DeviceMixin for the `SharedCapacityConsumed` field. Then we could have two separate fields on the device struct for referencing them, and only the mixin with the `SharedCapacityConsumed` field would need the name of the capacity pool. On the other hand, adding a third mixin type isn't necessarily an improvement.

mortent commented:
@johnbelamaric @klueska I've updated the PR to use the abstract pools design. The large example of the NVIDIA GPU is not updated yet, as I want to wait until we have (or at least close to) agreement on the API before I do the work on that. Please take a look.

klueska commented:
> The large example of the NVIDIA GPU is not updated yet, as I want to wait until we have (or at least close to) agreement on the API before I do the work on that.

Note, that example was generated from this:
https://github.com/kubernetes-sigs/wg-device-management/pull/38

pohly commented:
> I left the composite device in there, since I don't think we have fully decided on the path forward there, it is a relatively simple thing to remove, and we probably need it in the v1beta1 API anyway.

My conclusion after this week's WG Device Management meeting (also added to https://docs.google.com/document/d/1sCfO4hJUUhZpzld-_wEZi85XqSclEVvjUfsEe2-duZs/edit?tab=t.0):

The current consensus (2024-02-04) for DRA is:

* Add new fields as needed to v1beta1 BasicDevice.
* In v1beta2, eliminate the one-of in Device and move all those fields in BasicDevice up one level into Device.
* Mitigate issues with clients ignoring fields by ensuring that before promotion of a feature to beta, all relevant clients from previous releases know about the fields being promoted and fail in a sensible way when the corresponding feature is disabled in those older clients.

"Fail in sensible way" in this case means that the scheduler must ignore devices which depend on partitioning when `DRAPartitionalDevices` is turned off.

We still have a problem when a DRA driver creates a ResourceSlice with fields set for partitionable devices and the API server silently strips the fields.

I'll let you and @johnbelamaric figure out how to deal with that and how to describe these failure modes in the PRR section (version skew, feature disablement, etc.). I also left a comment in the doc: https://docs.google.com/document/d/1sCfO4hJUUhZpzld-_wEZi85XqSclEVvjUfsEe2-duZs/edit?disco=AAABdXllT24 might be a better place to discuss because it has threading and more context.

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,470 @@ non-goals of this KEP.
+  // +optional
+  Capacity map[QualifiedName]DeviceCapacity
+
+  // SharedCapacityConsumed defines a list of references to
+  // shared capacities defined in the same ResourceSlice. Each
+  // of the references points to a named shared capacity and
+  // includes the quantity of that capacity that the devices
+  // using this mixin will need.
+  //
+  // During allocation, a device can only be allocated if
+  // sufficient shared capacity is available to meet the
+  // capacity needed by all the references provided.
+  //
+  // Any capacity references here will be overridden by references
+  // defined in the device.
+  //
+  // The maximum number of consumed capacities is 32.
+  //
+  // +optional
+  SharedCapacityConsumed []MixinCapacityConsumedRef 
```
I don't like the idea of a mixin being able to define capacity comsumption. I explicitly disallowed this in my design because it hides too many of the details when you are looking at the device definition itself (or now the pool definition). Mixins are useful for declaring the set of capacities that should be made available by a pool, but it should be the pool (or device) itself that defines what (other) pool a given capacity is pulled from.

mortent replied:
So I think mixins in general hides some detail, since understanding the device/pool requires looking at other objects. And I think the more of these constructs we add, the more cognitive effort is required to understand what is going on. The tradeoff is obviously that fewer of these constructs means more duplication.

I'm not sure if I agree that capacity consumption on its own is worse in mixins than attributes and capacity. The way this design allows mixins to define the consumed capacity without specifying the pool makes things a little more complicated, but on the other hand I think the way it allows users to define profiles (to use the NVIDIA terminology) for partitions that can be reused across pools can also make it easier to understand if used wisely. But misuse of the mixin constructs from this design certainly makes it possible to create structures that will be hard to understand.

klueska replied:
Thinking about it more, I'm not even sure what it means for a mixin to include consumed capacity. By definition, all capacity must ultimately be pulled from some concrete device (or pool in the new design). So are you saying that mixins can actually reference a concrete pool, that itself may include some other set of mixins? Can you provide a simple example here in the comments where this becomes particularly useful (and without it it becomes overly verbose)?

mortent commented:
> I'll let you and @johnbelamaric figure out how to deal with that and how to describe these failure modes in the PRR section (version skew, feature disablement, etc.). I also left a comment in the doc: https://docs.google.com/document/d/1sCfO4hJUUhZpzld-_wEZi85XqSclEVvjUfsEe2-duZs/edit?disco=AAABdXllT24 might be a better place to discuss because it has threading and more context.

Ack. Thinking about this now, it isn't obvious to me how we should do this. I think this is an issue we need to solve regardless of whether we use abstract pools or hierarchy of devices.

johnbelamaric commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,470 @@ non-goals of this KEP.
+   partition will specify how much capacity it will consume from the capacity
+   pool, which allows the scheduler to keep track of available capacity in
+   a pool and decide whether a partition can be allocated. A device will
+   separatly set the capacities that is published for the partition.
```
s/separatly/separately/

s/set the capacities that is published/define the capacities that are accessible to the user/

johnbelamaric commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,470 @@ non-goals of this KEP.
+1. The `CapacityPoolMixins` field defines a list of named `CapacityPoolMixin`s. These
+   define a set of capacities that can be used to extend the capacities explicitly
+   defined in a `CapacityPool`. This allows for reduced duplication if there are many
+   identical physical devices that must be represented as capacity pools. Mixins
```
s/Mixins/CapacityPoolMixins/

johnbelamaric commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,470 @@ non-goals of this KEP.
+   define a set of capacities that can be used to extend the capacities explicitly
+   defined in a `CapacityPool`. This allows for reduced duplication if there are many
+   identical physical devices that must be represented as capacity pools. Mixins
+   can not be refrenced directly by devices.
```
s/refrenced/referenced/

johnbelamaric commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,470 @@ non-goals of this KEP.
+
+1. The `Includes` field serves to reference a set of `CompositeDeviceMixin`s that a
+   `CompositeDevice` can reference to extend the set of attributes,
+   capacities and consumed capacity it definces explicitly. If the referenced
```
s/definces/defines/

johnbelamaric commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,470 @@ non-goals of this KEP.
+  //
+  // The names of the pools must be unique in the ResourceSlice.
+  //
+  // The maximum number of pools is 32.
```
So in the NVIDIA case, this limits us to 32 physical GPUs in a ResourceSlice. Which is OK, since you can have more than one slice per node, if we get to very very large nodes.

mortent replied:
Yeah, it can probably be discussed if these are the right numbers, but in this case users can always spread things across multiple `ResourceSlices`. Each NVIDIA GPU (ref the A100 example) includes a relatively large number of devices, each of which has quite a few annotations and capacities. So I think there might be many reasons for splitting it up into multiple ResourceSlice objects if there are a lot of GPUs.

klueska replied:
For MIG there are 25 possible MIG devices + 1 full GPU so 26*8=208 total devices for a DGXA100 machine, for example. This is already spilling into more than one slice due to the 128 device limit. With the initial design, my plan was to create one slice per GPU and its MIG (so 8 slices total with 26 devices on each). The one downside being that I need to duplicate all of the mixins in each slice since they cannot span slices.

johnbelamaric commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,470 @@ non-goals of this KEP.
+  // Conflicting capacities from those provided via other mixins are
+  // overwritten by the ones provided here.
+  //
+  // The maximum number of attributes and capacities combined is 32.
```
I do worry 32 is too low. How many are on, say, a MIG partition today?

mortent replied:
So for the top-level partition (gpu-0) in the NVIDIA A100 example there are 12 attributes and 13 capacities. Note that this includes capacities like `memorySlice0...memorySlice7`. So a total of 25, which is kinda close to the max number, but also maybe not quite what it would look like for a real device.

This number is not changes in this PR, so maybe @klueska has some input here.

johnbelamaric replied:
@klueska had an "intrange" resource type concept in an early prototype. So rather than listing out memorySlice0..7, you would have a single intRange capacity [0..7], and your consumers would look like [0], [1], [0..1], etc.

This would be super useful to compact the inferentia example, too. And maybe even the TPU one if we could do, [0,1], [0,2] (ie, tuples not just ranges).

But parsing that as a resource.Quantity would be a major change since that's used all over.

mortent replied:
Ah, I was thinking that exposing the actual memory slices as capacities might not really be necessary? But I might be wrong here. What is the use-case for that?

johnbelamaric replied:
I don't think they need to be exposed to the user as capacities, but they are needed for the shared consumption for sure. I do know though that external controllers may want to "pin down" to a particular partition, if they don't want to let the scheduler pick the partition. I think other attributes probably service that purpose (for example, partition name), but I can imagine someone using the memorySliceX in a CEL expression for that purpose.

johnbelamaric commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,470 @@ non-goals of this KEP.
-the scheduler decides to allocate it.
+to include a set of mixins in a device or capacity pool definition, and (2) the
+ability to express that multiple devices draw from the same pool of capacity, so
+allocation of one device might make other device unallocatable.
```
s/other device/other devices/

johnbelamaric commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -549,105 +802,95 @@ devices:
-The scheduler can then allocate such devices using the same algorithm it uses
-for `Basic` devices.
+"flatten" them into a set devices that pull capacity from a set of pools. The
+scheduler can use this information to allocate partitionable devices without
```
s/partitionable devices/device partitions/

johnbelamaric commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -549,105 +802,95 @@ devices:
+When the scheduler tries to allocate devices for a new ResourceClaim, it will
+go through the list of devices and check which of them are available. Since it
+knows the free capacity in all capacity pools, it can easily check whether a
+specific device can be allocated.
```
Add something like:

Note that since the scheduler will go through these in order and select the *first* (not *best*) fit, it is recommended that `ResourceSlice`s include devices in smallest-to-largest order. Otherwise, the largest one will always be allocated. This may be improved through scoring mechanisms, but that is outside of the scope of this KEP.

And maybe link to the scoring issues.

johnbelamaric replied:
(As an aside, I suppose in theory we could add something in the claim that let the user specify the sort order to use here, rather than just taking what is in the slice. But that sounds...hard and confusing.)

johnbelamaric commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -663,126 +906,116 @@ together in a discrete set of ways to construct a MIG device with a particular
+a set of devices that pull specific capacity from that pool.
 
-For example, an unrolled "source" device for an NVIDIA A100 GPU looks as follows:
+For example, capacity pool for an NVIDIA A100 GPU looks as follows:
```
s/, capacity/, a capacity/

johnbelamaric reviewed (commented):
Looking good

mortent commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,470 @@ non-goals of this KEP.
+  //
+  // +optional
+  // listType=atomic
+  SharedCapacityPools []CapacityPool
```
I'm not really sure if the `Shared` prefix here adds any value. I think maybe we should just remove it.

johnbelamaric replied:
Yeah, it's iffy. But I want to differentiate between this and [per-device allocatable resources](https://github.com/kubernetes/enhancements/pull/5104#discussion_r1943844212), which are for user consumption.

mortent replied:
I did remove the `Shared` prefix in the updated design. I think it works pretty well without it, but let me know if you disagree.

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -490,17 +738,19 @@ deviceMixins:
-  composite:
-    capacity:
-      memory:
-        quantity: 40Gi
+  capacity:
+    memory: 40Gi
+  sharedCapacityConsumed:
+  - name: memory
+    capacity: 40Gi
```
I'm not sure I understand this:
```
- name: common-gpu-capacities
  capacity:
    memory: 40Gi
  sharedCapacityConsumed:
  - name: memory
    capacity: 40Gi
```

What is the distinction between the `capacity` of 40GB of memory and the `sharedCapacityConsumed` of 40GB of memory? Is this to support the case where some capacity might be local to the mixin, but other capacity comes from elsewhere? What does it even mean for capacity to be local to a mixin?

mortent replied:
The idea is to separate the capacity that is consumed from the shared pool from what is advertised as capacity available to workloads. So in these scenarios, we would expect the `sharedCapacityConsumed` to be larger than the `capacity`. And they obviously don't necessarily have to have matching names.

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -490,17 +738,19 @@ deviceMixins:
+  - name: memory
+    capacity: 40Gi
 devices:
 - name: gpu-0
   composite:
     includes:
     - name: system-attributes
     - name: common-gpu-attributes
     - name: common-gpu-capacities
+      capacityPool: gpu-0-pool
```
Just having `capacityPool` tacked on here like this feels unreadable to me. Is this saying that any capacities listed in the `sharedCapacityConsumed ` section of the `common-gpu-capacities` mixin should be pulled from `gpu-0-pool`?

mortent replied:
Correct. I do agree that it reduces readability somewhat, but it makes it possible to reuse device mixins across different devices. I think the example in ["Extending device and capacity pool with a set of mixins"](https://github.com/mortent/enhancements/tree/ClarifyPartitionableDevices/keps/sig-scheduling/4815-dra-partitionable-devices#extending-device-and-capacity-pool-with-a-set-of-mixins) is good example of where it is useful, as we can use the same `common-gpu-capacities` mixin for all the different devices, even though they belong to different pools.

I think this is related to your [previous comment](https://github.com/kubernetes/enhancements/pull/5066#discussion_r1944615581) in that I think we either should allow `sharedCapacityConsumed` in device mixins with the ability to provide the capacity pool when it is referenced from the device (so what is in the current proposal), or we don't allow `sharedCapacityConsumed` in mixins at all. I think it is less useful to have them in a mixin if that also ties the mixin to a specific capacity pool.

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -663,126 +912,116 @@ together in a discrete set of ways to construct a MIG device with a particular
+sharedCapacityPools:
+- name: gpu-0-pool
+  ...
+deviceMixins:
+- name: mig-1g.5gb
+  capacity:
+    copy-engines: "1"
+    decoders: "0"
+    encoders: "0"
+    jpeg-engines: "0"
+    memory: 4864Mi
+    memorySlice0: "1"
+    multiprocessors: "14"
+    ofa-engines: "0"
+  sharedCapacityConsumed:
+  - name: copy-engines
+    capacity: "1"
+  - name: decoders
+    capacity: "0"
+  - name: encoders
+    capacity: "0"
+  - name: jpeg-engines
+    capacity: "0"
+  - name: memory
+    capacity: 4864Mi
+  - name: memorySlice0
+    capacity: "1"
+  - name: multiprocessors
+    capacity: "14"
+  - name: ofa-engines
+    capacity: "0"
```
I still can't wrap my head around what it means for a mixin to consume shared capacity from somewhere. Is it not enough to simply list the capacities in the mixing and then when you include a reference to a pool at the point you include the mixin in a deivce, that indicates which pool that capacity is pulled from.

mortent replied:
Closing this since we have updated this part of the design.

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,470 @@ non-goals of this KEP.
+  // the device.
+  //
+  // +required
+  Capacity resource.Quantity
```
Should this be a DeviceCapacity (to match what is held in an actual CapacityPool) and not just a resource.Quantity?

mortent replied:
Yeah, I missed this.

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,470 @@ non-goals of this KEP.
+  // the device.
+  //
+  // +required
+  Capacity resource.Quantity
```
Should this be a DeviceCapacity (to match what is held in an actual CapacityPool) and not just a resource.Quantity?

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,470 @@ non-goals of this KEP.
+  // The maximum number of attributes and capacities combined is 32.
+  //
+  // +optional
+  Capacity map[QualifiedName]DeviceCapacity
```
As I mentiond in another comment. I don't think devices should have the ability to define their own intrinsic capacities. *All* capacity of a device should be pulled from (one or more) pools.

In my mental model with this new pool design, there would be AttributeMixins and CapacityMixins. Devices can  include AttributeMixins and CapacityMixins, but ResourcePools can *only* include CapacityMixins.

When constructing a device you then include its set of desired attribute mixins, and then (in a separate field) list the set of CapacityMixins you want attributed to the device -- each of which *must* including a reference to a pool where the capacities listed in that mixin come from.

johnbelamaric replied:
I think that the internal capacity consumed by partitions being the same as the values advertised to consumers is just a coincidence of the MIG case. I don't think that will be true in the general case and I don't want to lock us into it. For example, I can imagine a device that needs some overhead associated with the internal allocation before making the resource available to consumers. Or, even in the MIG case, telling the external consumers about "memory slices" seems odd.

Maybe there's a way to build the API to simplify the MIG case but still allow the general case.

johnbelamaric replied:
Or, someone may build a device that advertises logical resources but consumes physical ones: I can service 1000 requests per second if I consume 1 core, or 2000 requests per second if I use 2 cores. So, internally they are consuming "cores" externally they are advertising "requests per second".

mortent replied:
So with the discussion offline, I think we are converging on something like:

* DeviceMixin will include both attributes and capacity. Capacity means the advertised resource availability of a device. These can only be referenced from devices and they are referenced by providing only the name.
* CapacityPoolMixin will include only capacity. These can only be referenced from CapacityPools
* SharedCapacityConsumedMixin (we should find a better name) includes the shared capacity consumed, but does not specify a capacity pool. They can only be referenced from devices. Both the name and the capacity pool will need to be provided when they are included.

I will try to get the PR updated tonight.

mortent replied:
PR has been updated.

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -272,10 +272,10 @@ On a high level, this proposal is based around two new concepts:
-   that devices can draw from. `CapacityPoolMixins` and `DeviceMixins` defines mixins
-   for capacity pools and devices.
+   that devices can draw from. `CapacityPoolMixins`, `DeviceMixins`, and
+   `DeviceCapacityConsumptionMixins` defines mixins for capacity pools and devices.
```
```suggestion
   `DeviceCapacityConsumptionMixins` define mixins for capacity pools and devices.
```

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -272,10 +272,10 @@ On a high level, this proposal is based around two new concepts:
+   that devices can draw from. `CapacityPoolMixins`, `DeviceMixins`, and
+   `DeviceCapacityConsumptionMixins` defines mixins for capacity pools and devices.
 
 1. The `SharedCapacityPools` field is a list of named `CapacityPool`s. Each
```
```suggestion
1. The `CapacityPools` field is a list of named `CapacityPool`s. Each
```

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,461 @@ non-goals of this KEP.
+   object that defines some properties and that when used, will extend the properties
+   they define explicitly on the device or resource pool. The objective is to reduce
+   the duplication and overall footprint of `ResourceSlices`.
+
```
```suggestion
```

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,461 @@ non-goals of this KEP.
+  //
+  // +required
+  // +listType=atomic
+  ConsumesCapacity []CapacityConsumptionRef
```
```suggestion
  ConsumedCapacity []DeviceCapacityConsumptionRef
```

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
````diff
@@ -252,222 +252,461 @@ non-goals of this KEP.
 ```yaml
+capacityPoolMixins:
+- name: gpu-pool-mixin
+  capacity:
+    memory: 40Gi
````
I believe this should be (all throughout):
```
  capacity:
    memory:
      value: 40Gi
```

mortent replied:
I've cleaned up the updated examples and will make sure do it this way for the rest of the examples.

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,461 @@ non-goals of this KEP.
+  // The maximum number of capacities is 32
+  //
+  // +required
+  CapacityConsumption map[QualifiedName]DeviceCapacity
```
I think we can just call this `capacity` -- its clear from where its used, what the context is.

mortent replied:
Yeah, agree. I think the name of this field should be the same in both the mixin and the device (i.e. the `DeviceCapacityConsumption` struct) should match, so I've updated both.

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,461 @@ non-goals of this KEP.
+  // The maximum number of attributes and capacities combined is 32.
+  //
+  // +optional
+  Capacity map[QualifiedName]DeviceCapacity
```
Does it make sense to rename this `AdvertisedCapacity` to contrast it with `ConsumedCapacity`?

mortent replied:
Yeah, I did think about a more descriptive name here. I think it makes sense, but I am not sure if this makes the mapping from devices in the `ResourceSlice` to the properties of the CEL expressions in the `ResourceClaims` less obvious? 
I was also uncertain about introducing differences in naming between `BasicDevice` and `CompositeDevice`, although that would be something we can fix in `v1beta2`.

thockin replied:
My $0.02 - it's fine as is.

bowei commented:
/cc @bowei

mortent commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,461 @@ non-goals of this KEP.
+  // are also defined in any of the referenced mixins.
+  //
+  // +optional
+  CapacityConsumption DeviceCapacity
```
Yeah, I'm not sure what I was thinking when I named the field `Mixins`... But fixed now.

mortent commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,461 @@ non-goals of this KEP.
+
+// DeviceCapacityConsumptionRef defines a set of capacities that
+// a device will consume from a capacity pool.
+type DeviceCapacityConsumptionRef struct {
```
Yeah, agree that it could have a better name. Since this specifies the capacity consumption for a device, I think just removing the `Ref` suffix works pretty well. So it is updated to just be `DeviceCapacityConsumption` now.

mortent commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,465 @@ non-goals of this KEP.
+  //
+  // +optional
+  // +listType=atomic
+  Includes []CapacityPoolMixinRef
```
@klueska Is there a particular reason why the original design has references as a list of objects with just the `Name` field rather than just a list of strings? I've kept all the `<Something>Ref` types consistent with this approach, but using a separate struct is a little bit more verbose than just a list of strings.

klueska replied:
It was just to leave it open to extension in the future. We do the same for the reference to rsourceclaims in a pod spec, for example.

mortent replied:
Yeah, that makes sense.

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,465 @@ non-goals of this KEP.
+  //
+  // +required
+  // +listType=atomic
+  ConsumesCapacity []DeviceCapacityConsumption
```
This is just a suggestion, I forget the convention here

mortent replied:
I don't think we have a convention here. The other fields with similar names that were present in earlier versions of the design have been renamed. I don't feel strongly about this, but I think `ConsumesCapacity` is slightly better as the field describes the capacity the device will consume once allocated.

klueska replied:
@johnbelamaric @thockin thoughts here? I remember some discussion on conventions around verbs vs nouns, past vs. present tense, etc.

thockin replied:
I don't recall a strong convention.  A quick survey shows the past-tense is probably more consistent, but I am open to counter-envidence.

klueska replied:
@mortent let's change this to `consumed`, as i also recall past-tense being the more common pattern here.

mortent commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,465 @@ non-goals of this KEP.
+  // The maximum number of capacities is 32.
+  //
+  // +optional
+  Capacity DeviceCapacity
```
Should we also enforce that the "effective" number of capacities for a specific device (i.e. the number of capacities defined directly plus any included through mixins) also remain below the max number? I think we probably should, but just wanted to see if anyone has an opinion on this.

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,465 @@ non-goals of this KEP.
+  // The maximum number of capacities is 32.
+  //
+  // +optional
+  Capacity DeviceCapacity
```
I had suggested making this a `map[QualifiedName]DeviceCapacity` (which seems to have gotten lot). Is there a reason not to do this to be symmetrical with other types?

mortent replied:
Sorry, I missed that one. Yeah, it should be a map. Updated now.

pohly commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,465 @@ non-goals of this KEP.
+ResourceSlice API:
+
+1. It introduces four new fields on the `ResourceSliceSpec`. First, `CapacityPools`
+   defines a list of `CapacityPools`, each of which defines a set of capacities
+   that devices can draw from. `CapacityPoolMixins`, `DeviceMixins`, and
```
Nit:

```suggestion
   defines a list of `CapacityPools`. `CapacityPoolMixins`, `DeviceMixins`, and
```

Details about `CapacityPools` are available below. Having one place to learn about them seems better to me.

pohly commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,465 @@ non-goals of this KEP.
+  //
+  // The names of the pools must be unique in the ResourceSlice.
+  //
+  // The maximum number of pools is 32.
```
The limit for the other new items (mixins, composite devices) is defined so that the overall number is the same as the before for basic device, so the maximum size should be about the same as before.

This here is allowed to add something on top of that. Each pool is small, so this should be fine.

I'm mostly just wondering how we can be a bit more sure that it's really okay. Doing a back-of-the-envelope calculation as in the past is getting more difficult. Perhaps we should have a validation unit test which creates a slice with all fields at their maximum and then checks the size of protobuf encoding?

I had that idea before, but I don't remember whether I implemented it... :sweat_smile: 

There's no need to write to etcd, [`ResourceSlice.Size`](https://github.com/kubernetes/kubernetes/blob/9e555875e79f60a238952e91e91f79f32c053f9c/staging/src/k8s.io/api/resource/v1beta1/generated.pb.go#L3340) accurately predicts the size.

This could be added to the "unit tests" section of this KEP.

mortent replied:
I couldn't find any test like this in the code base, so added a short section to the "unit tests" section describing that we plan to add this.

pohly commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -508,6 +755,10 @@ devices:
         int: 0
       uuid:
         string: GPU-4cf8db2d-06c0-7d70-1a51-e59b25b2c16c
+    consumesCapacity:
```
Not really about this line...

Each of the GPUs is shown with attributes "index" and "minor". I think having just one per-device attribute (the "uuid") is sufficient and easier to understand.

As it stands now, as I reader I am trying to understand what "index" and "minor" mean although I shouldn't bother yet.

klueska replied:
yes, for the example, no need to overcomplicate it I guess...

pohly commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -547,107 +806,126 @@ devices:
+  capacity:
+    memory: 
+      value: 10Gi
+deviceCapacityConsuptionMixins:
```
```suggestion
deviceCapacityConsumptionMixins:
```

pohly commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
````diff
@@ -792,15 +1060,19 @@ allocated together with the third one because they compete on the capacity
-out their `ConsumesCapacityFrom` fields appropriately.
-
 ```yaml
+******* WILL BE UPDATED ONCE WE MORE FEEDBACK ON THE DESIGN *******
````
Ready to be updated?

mrunalp commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,465 @@ non-goals of this KEP.
+  //
+  // +optional
+  // +listType=atomic
+  DeviceMixins []DeviceMixin
+
+  // DeviceCapacityConsumptionMixins represents a list of capacity
+  // consumption mixins, each of which contains a set of capacities
+  // that a device will consume from a capacity pool.
+  //
+  // This makes it possible to define a set of shared capacities that
+  // are not tied to a specific pool. The pool is inferred by context
+  // in which the DeviceCapacityConsumptionMixin is referenced from
+  // the device.
+  //
+  // The total number of device mixins, device capacity consumption mixins, 
+  // capacity pool mixins, basic devices, and composite devices must be
+  // less than 128.
+  //
+  // +optional
+  // +listType=atomic
+  DeviceCapacityConsumptionMixins []DeviceCapacityConsumptionMixin
+
+  // CapacityPoolMixins represents a list of capacity pool mixins, i.e.
+  // a collection of capacities that a CapacityPool can "include"
+  // to extend the set of capacities it already defines.
+  //
+  // The main purposes of these mixins is to reduce the memory footprint
+  // of capacity pools since they can reference the mixins provided here rather
+  // than duplicate them.
+  //
+  // The total number of device mixins, device capacity consumption mixins, 
+  // capacity pool mixins, basic devices, and composite devices must be
+  // less than 128.
+  //
+  // +optional
+  // +listType=atomic
+  CapacityPoolMixins []CapacityPoolMixin
```
Yeah, I like this nesting :+1:

johnbelamaric replied:
me too

klueska replied:
What about even:
```
mixins:
  device: 
    attribute: ...
    capacity: ...
    capacityConsumption: ...
  capacityPool::
    capacity: ...
```
That way we don't need separate mixin types for `DeviceMixins`, `DeviceCapacityConsumptionMixin`, and `CapacityPoolMixin`. We just have `AttributeMixin` and `CapacityMixin` and they get used where appropriate

johnbelamaric replied:
works for me. @mortent will be back tomorrow

mortent replied:
So I think the initial proposal here to nest all mixins under a single Mixins structs is a clear improvement.

@klueska My assumption in the second suggestion is that each of the `attribute`, `capacity`, `capacityConsumption` fields within the device struct would be lists of named mixins. So more expanded it would look something like:
```
mixins:
  device: 
    attribute:
    - name: device-attributes-mixin
      attributes:
        minor:
          int: 3
    capacity:
    - name: device-advertised-capacities-mixin
      capacity:
        memory:
          value: 40Gi
    capacityConsumption:
    - name: device-consumed-capacities-mixin
      capacity:
        memory:
          value: 40Gi
  capacityPool::
    capacity:
    - name: capacity-pool-mixin
      capacity:
        memory:
          value: 40Gi
```

If this is the suggestion, I'm less convinced about this one for a few reasons:
* We end up with nested `attribute.attributes` and `capacity.capacity` fields, which feels a bit awkward.
* Re-using a `CapacityMixin` type saves us some duplication, but we loose some flexibility to evolve the mixins for capacity pools and devices separately in the future.
* We end up with four different mixin types, so a device will need three different `Includes` fields (`IncludesAttributes`, `IncludesCapacities`, and the `Includes` field nested under `consumesCapacity`.

klueska replied:
Yeah, that second suggestion looks awkward. Let's stick with the single level of nesting

thockin commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,465 @@ non-goals of this KEP.
+  // consume from those pools.
+  //
+  // The capacities can be defined either by referencing one
+  // or more DeviceCapacityConsumptionMixins for by listing
```
"...or by listing"  (says "for")

thockin commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -547,107 +806,126 @@ devices:
-"flatten" them into something that looks exactly the same as a `Basic` device.
-The scheduler can then allocate such devices using the same algorithm it uses
-for `Basic` devices.
+"flatten" them into a set devices that pull capacity from a set of pools. The
```
Can we include something like a kubectl cmd (not sure where it should go) or even a non-default plugin which does the flatten operation and prints the flattened thing?

Maybe we actually want a generic verb for this which fetches related things and prints them all together?  `kubectl flatten gateway foo`  

@kubernetes/sig-cli-leads

pohly replied:
We have a similar concern over in https://github.com/kubernetes/enhancements/pull/5034: I definitely see a need to improve the CLI to help users make sense of the information that is available because it's not all in one object or even a single type. But not for alpha, and perhaps even as a separate KEP if it's considered non-blocking for graduation of the functionality itself.

thockin reviewed (commented):
Maybe surprising, but I don't hate it. :)

It's definitely hard to think about, so having some CLI to expand it all seems necessary.

One thing that might go well in this doc is a two-column view of a fully-flattened device and an equivalently mixin'ed device.  I understand that would be tedious to do.  Maybe add two YAML files in this repo and put the `diff -u` of them in the KEP in a ` ```diff ` block?

pohly commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,465 @@ non-goals of this KEP.
+  // The maximum number of pools is 32.
+  //
+  // +optional
+  // listType=atomic
```
For me, the deciding factor is whether this slice has one owner or many and how it gets updated.

When using `listType=map`, the tracking information for who owns which entry explodes. This is unnecessary here because there is a single owner, the DRA driver. There's also no need for patching specific entries.

thockin replied:
Right, it was the `managedFields` argument.  This is more a note for the declarative validation effort than  for this KEP.

klueska commented:
> One thing that might go well in this doc is a two-column view of a fully-flattened device and an equivalently mixin'ed device. I understand that would be tedious to do. Maybe add two YAML files in this repo and put the `diff -u` of them in the KEP in a `` ```diff `` block?

@thockin I had put this together for the original KEP in 1.32:
https://github.com/kubernetes-sigs/wg-device-management/pull/38

When ran, it generates all of the devices necessary to represent full GPUs and their MIG partitions on a mock DGXA100 machine. It then prints both the actual slice with its mixins as well as the "flattened" slice. 

This is how the *long* example for MIG at the end of the KEP was generated.

It's probably worthwhile to update this to the new abstractions being proposed in this KEP and update that example in the process.

alculquicondor reviewed (commented):
It looks like the contentious parts are more about the API than the scheduling algorithm, so I'm going to approve for SIG Scheduling in advance.

/approve

thockin commented:
I guess I am OK with the API.  All my comments are minor.

/approve
/lgtm

mortent commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,465 @@ non-goals of this KEP.
+   called `ConsumesCapacity` and defines the capacity the device will draw
+   from the capacity pools.
+
+1. The `Includes` field serves to reference a set of `DeviceMixin`s that a
```
Agree. I've fixed this and also used a similar structure underneath the `Mixins` field added based on the suggestion in https://github.com/kubernetes/enhancements/pull/5066#discussion_r1948953578

mortent commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,465 @@ non-goals of this KEP.
+The implementation of these concepts requires several changes to the
+ResourceSlice API:
+
+1. It introduces four new fields on the `ResourceSliceSpec`. First, `CapacityPools`
```
Fixed.

mortent commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,465 @@ non-goals of this KEP.
+  // Conflicting attributes from those provided via other mixins are
+  // overwritten by the ones provided here.
+  //
+  // The maximum number of attributes and capacities combined is 32.
```
Good question. I think the primary concern we try to address is the size limit in etcd, but that doesn't really apply to the fully unpacked device (since we never write that to etcd). I'm not sure if it might impact the performance of evaluating the CEL expressions, but I wouldn't think so.

pohly replied:
I think we have more flexibility for runtime costs and thus don't need a limit for the unpacked device. If someone wants to use this feature for something complex, they have to be aware that evaluation will also become slower.

mortent commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,465 @@ non-goals of this KEP.
+  // Name refers to the name of a device mixin in the pool.
+  //
+  // +required
+  Name string
```
I don't think we have any plans that would need it, but doing it this way does allow us to add it if necessary, ref https://github.com/kubernetes/enhancements/pull/5066#discussion_r1946964462. It might be that we should just use a list of strings here instead to keep it as simple as possible and then just handle it if/when we do have an actual need for any additional fields.

mortent commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -252,222 +252,465 @@ non-goals of this KEP.
+  // Attributes defines the set of attributes for this mixin.
+  // The name of each attribute must be unique in that set.
+  //
+  // To ensure this uniqueness, attributes defined by the vendor
```
Yeah, we are doing validation in similar situations today and plan to add validation for this.

mortent commented:
@klueska I've updated the KEP with the changes to the mixins and I have updated the large example.

@pohly I've also adressed your other comments. If this looks good to you, can you clear the "changes requested" status?

pohly commented:
/lgtm

I'm keeping the hold for now because:
* @klueska might also want to have another look.
* I think it would be good to squash before lifting the hold.

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -0,0 +1,2899 @@
+
+## Proposal
+
+On a high level, this proposal is based around two new concepts:
```
```suggestion
At a high level, this proposal is based around two new concepts:
```

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -0,0 +1,2899 @@
+On a high level, this proposal is based around two new concepts:
+
+1. Capacity pool, which is a construct for defining the capacities available for
+   sharing among partitions of a single physical device. Drivers/controllers will
```
```suggestion
   sharing among partitions of a single physical device. Drivers will
```

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -0,0 +1,2899 @@
+    1. The `CapacityPool` field defines a list of named `CapacityPoolMixin`s. These
+      define a set of capacities that can be used to extend the capacities explicitly
+      defined in a `CapacityPool`. This allows for reduced duplication if there are many
+      identical physical devices that must be represented as capacity pools. CapacityPoolMixins
```
```suggestion
      identical physical devices that must be represented as capacity pools. `CapacityPoolMixin`s
```

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -0,0 +1,2899 @@
+
+    1. The `Device` field is a list of named `DeviceMixin`s. These define
+      attributes and capacities that can be used to extend what is defined
+      explicitly in a `CompositeDevice`. Mixins can not be allocated directly,
```
This is the first time that `CompositeDevice` is mentioned in the revised version of this KEP.#

As such, let's reword this to the following:
```suggestion
      explicitly in a new device type called `CompositeDevice` (introduced in more detail below). Mixins cannot be allocated directly,
```

pohly replied:
Side note: for now in this KEP revision we assume that we keep `CompositeDevice`. That's correct, no need to change anything.

In https://docs.google.com/document/d/1sCfO4hJUUhZpzld-_wEZi85XqSclEVvjUfsEe2-duZs/edit?tab=t.0 I proposed some changes that would make it safe to flatten the `Device` struct. This is something that we can try to describe in the next KEP update (https://github.com/kubernetes/enhancements/pull/5069), or treat it as an implementation choice and retro-actively update the KEP after we have agreed on the implementation in https://github.com/kubernetes/kubernetes/pull/129970. Given the time constraints I am leaning towards the latter.

klueska commented on keps/sig-scheduling/4815-dra-partitionable-devices/README.md:
```diff
@@ -0,0 +1,2899 @@
+defines the capacity the device draws from the capaity pool. This allows the
+definition of each partition to be quite compact.
+
+After all mixins have been "flattened", the scheduler end up with a list
```
```suggestion
After all mixins have been "flattened", the scheduler ends up with a list
```

klueska reviewed (commented):
Looking really good. Thanks for all the iterations on this @mortent. Just a few minor comments.

mortent commented:
@klueska @pohly I have addressed the comments and squashed the commits.

klueska reviewed (commented):
/lgtm
/unhold
